homeassistant/components/light/mqtt/__init__.py

```diff
@@ -0,0 +1,72 @@
+"""
+Support for MQTT lights.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/light.mqtt/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import light
+from homeassistant.components.mqtt import ATTR_DISCOVERY_HASH
+from homeassistant.components.mqtt.discovery import MQTT_DISCOVERY_NEW
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import HomeAssistantType, ConfigType
+
+from . import schema_basic
+from . import schema_json
+from . import schema_template
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['mqtt']
+
+CONF_SCHEMA = 'schema'
+
+
+def validate_mqtt_light(value):
+    """Validate MQTT light schema."""
+    schemas = {
+        'basic': schema_basic.PLATFORM_SCHEMA_BASIC,
+        'json': schema_json.PLATFORM_SCHEMA_JSON,
+        'template': schema_template.PLATFORM_SCHEMA_TEMPLATE,
+    }
+    return schemas[value[CONF_SCHEMA]](value)
+
+
+PLATFORM_SCHEMA = vol.All(vol.Schema({
```

Comment:
Instead of this, use `vol.Any(schema_template.PLATFORM_SCHEMA_TEMPLATE, schema_json.PLATFORM_SCHEMA_JSON, schema_basic.PLATFORM_SCHEMA_BASIC)` as `PLATFORM_SCHEMA`

Reply:
@balloob Problem with that is that it produces really bad validation errors (only the last error is logged). See https://github.com/home-assistant/home-assistant/pull/18227#discussion_r232449736 - but if you prefer the former that's ok too :)

Reply:
Maybe a reasonable way is to add a comment explaining the custom validator is added to improve error messages?